### PR TITLE
Adds tests for some missing coverage on dynamically created fields

### PR DESCRIPTION
**Pull request by colings86**
This adds some new tests to DocumentParserTests to make sure the DocumentParser behaves correctly when dynamically mapping fields. Especially testing that the dynamic setting works when dynamically mapping different field types. The gaps in the tests were mainly found using code coverage tools

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java [hunk 1]
```diff
@@ -283,4 +283,215 @@ public void testDynamicArrayWithTemplate() throws Exception {
...
+        DocumentMapperParser mapperParser = createIndex("test").mapperService().documentMapperParser();
+        String mapping = XContentFactory.jsonBuilder().startObject().startObject("type").field("dynamic", "strict")
+            .endObject().endObject().string();
+        DocumentMapper mapper = mapperParser.parse("type", new CompressedXContent(mapping));
+
+        BytesReference bytes = XContentFactory.jsonBuilder()
+                .startObject().startObject("foo")
+                    .field("bar", "baz")
+                .endObject().endObject().bytes();
+        expectThrows(StrictDynamicMappingException.class,
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java]: should we check the error message?

**Comment by jpountz** [comment]
Woohoo!!! I'm wondering if we should check the exception messages in the test (unless maybe the exception type is specific enouch?). Otherwise LGTM.

**Comment by rjernst** [comment]
I agree we should check the exception message.

**Comment by colings86** [comment]
@jpountz @rjernst I added a check on the exception message before I merged. Thanks both for the review

**Pull request merged**